Radial menu to remove PKA modkits with a right click from an empty hand.

GuillaumePrata opened the pull request:
## About The Pull Request
You can choose and remove one at a time, code blatantly stolen from the shotgun cabinet.
The crowbar functionality stays because Cyborgs need it to empty their PKA.
## Why It's Good For The Game
![Right click PKA](https://user-images.githubusercontent.com/55374212/204449791-209afc83-9e72-42c4-aa08-92c39216b979.png)

This gives more control for miners to swap modkits from one fight to the next or possibly mid fight.
Miners often have a favorite PKA setup that they use for most shifts, by making swapping easier we have a little more freedom to add modkits that are only useful occasionally (say a specific fauna counter) and these are more likely to be used with an easier swaps system.
The old way to remove modkits was just pain, it would remove every modkit at once and dump them on the floor.
Something something UX.

And hey, excuse to give better sprites to each modkit in the future. 
## Changelog
:cl: Guillaume Prata
qol: Right clicking a PKA with an empty hand will show a radial menu that lets you remove modkits one at the time.
/:cl:

GuillaumePrata committed:
Ported but not idea if it works

GuillaumePrata committed:
Merge remote-tracking branch 'upstream/master' into pka_right_click

GuillaumePrata committed:
Better var names

And it actually did work on the first commit, my face when my code work first try

OrionTheFox commented on code/game/objects/items/storage/boxes/job_boxes.dm:
```diff
@@ -59,10 +59,6 @@
 /obj/item/storage/box/survival/mining
 	mask_type = /obj/item/clothing/mask/gas/explorer/folded
 
-/obj/item/storage/box/survival/mining/PopulateContents()
-	..()
-	new /obj/item/crowbar/red(src)
-
```
Pleaseeee don't remove this, miners may deal with ruins that need doors pried open, may need to pry their way into a powerless outpost that has no staff to maintain it, basically need to open doors on a powerless planet. It's not solely for modkits.

Sealed101 replied:
why are you removing the crowbar

Serxule replied:
yeah miners have more uses for crowbars than just their PKA

Fikou reviewed (changes requested):
keep the crowbar

Fikou commented:
> > keep the crowbar
> 
> Why? It's a 50 cost iron tool that miners have access roundstart and get in the way of other items if we ever try things with survival boxes unless every future feature comes bundled with a "survival boxes are bigger too" change.

if such problems arise we can handle them at that point

GuillaumePrata commented:
Removing the crowbar also makes people ask why and likely learn that they can right click modkits out of the PKA if they look up this PR, changelog or another player explains directly.

If we leave the crowbar miners will just keep defaulting to how things were until they randomly learn about this PR.
And what do we get in return? Miners have crowbar access at roundstart, at most they will get locked on fire/airlocks between the shuttle and cargo if they late join.

GuillaumePrata commented:
For another angle, the crowbar was there to let miners remove modkits, something that is not necessary anymore as we have a better system.

So I still ask why we should keep the crowbar there?

Jackal-boop commented:
Just keep the crowbar. No reason to remove it

GuillaumePrata commented:
I already gave 3 reasons to remove it and no one gave a good reason to keep it.
Average mining PR experience.

Overuseddart commented:
While yes, miners CAN print the crowbar out round-start it doesn't mean that they should have too. A newer miner or miner that doesn't know about this change would likely go onto lava land and not be able to open a door which needs a crowbar, which is why the crowbar exists in the box, as well as being able to take out modkits. 

This change is good however, I would like to see miners making midnight mod kit changes to better suit their situation.

GuillaumePrata commented:
There are 2 crowbars on the shuttle and at least 1 on the outpost.
If miners forget a crowbar, walk back to the outpost, it's what curators have to do with when they go space exploring and forget tools.

Overuseddart commented:
The difference being that space exploring as a curator is optional, as your job is an RP one.
Newer players wouldn’t know where to get a crowbar and not know to bring one to lavaland until they reach an unpowered door and then not know how to get one as a miner as the auto lathe can be confusing to newer players.

GuillaumePrata commented:
There are 2 crowbars by the shuttle console, a toolbox on a table in the outpost, they can buy, print, ask, etc.

I'm all for helping with onboarding as mining sucks for a new player experience but I just don't think that walking to a ruin, realizing you can't go in and having to go back for tools is a bad experience for new miners.

It would actually make them more interested into returning later to see what is inside., we should have more ruins that need back and fourth planning like this instead of just strolling with no effort.

Overuseddart commented:
Thinking back on it walking to a ruin and not having the tools for it and going back to be more prepared is a good experience for newer miners. 
Maybe a system where you need specific things for ruins (as well as increasing the loot) and carried tools are limited would be fun. 

I’m onboard for this change now, it seems to be good.

Mothblocks commented on code/modules/projectiles/guns/energy/kinetic_accelerator.dm:
```diff
@@ -31,22 +31,62 @@
+/obj/item/gun/energy/recharge/kinetic_accelerator/attack_hand_secondary(mob/user, list/modifiers)
+	. = ..()
+	if(. == SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN)
+		return
```
```suggestion
		return SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN
```

Mothblocks commented on code/modules/projectiles/guns/energy/kinetic_accelerator.dm:
```diff
@@ -31,22 +31,62 @@
+		return
+
+	if(!LAZYLEN(modkits))
+		return
```
```suggestion
		return SECONDARY_ATTACK_CONTINUE_ATTACK_CHAIN
```

_secondary should never return `..()`, it'll call normal

Mothblocks commented on code/modules/projectiles/guns/energy/kinetic_accelerator.dm:
```diff
@@ -31,22 +31,62 @@
+
+	var/list/display_names = list()
+	var/list/items = list()
+	for(var/modkits_lenght in 1 to length(modkits))
```
Typo--length

Mothblocks commented on code/modules/projectiles/guns/energy/kinetic_accelerator.dm:
```diff
@@ -31,22 +31,62 @@
+		return
+	if(!user.put_in_hands(modkit_to_remove))
+		modkit_to_remove.forceMove(get_turf(src))
+	update_appearance()
```
Be explicit, such as `UPDATE_ICON`

Mothblocks commented on code/modules/projectiles/guns/energy/kinetic_accelerator.dm:
```diff
@@ -31,22 +31,62 @@
+	if(!istype(modkit_to_remove))
+		return
+	if(!user.put_in_hands(modkit_to_remove))
+		modkit_to_remove.forceMove(get_turf(src))
```
```suggestion
		modkit_to_remove.forceMove(drop_location())
```

GuillaumePrata replied:
`drop_location` alone was throwing errors so I checked how other items go around it and it was by declaring(?) a new `drop_location` before using it.

```
user = new(drop_location())			
M.forceMove(drop_location())
```

I have no idea if that is cleaner than `get_turf` so please tell me if I can do it in a better way.

Mothblocks replied:
I see plenty of working usages of it, what error were you getting?

GuillaumePrata replied:
I think it was an issue with undefined procs, I found it weird because there is another use of `drop_location` error on the file that was not throwing the same error.

Removed the line now and it is not throwing error so it was likely just me VSC or me being dumb.

Mothblocks commented on code/modules/projectiles/guns/energy/kinetic_accelerator.dm:
```diff
@@ -31,22 +31,62 @@
+		var/image/item_image = image(icon = thing.icon, icon_state = thing.icon_state)
+		if(length(thing.overlays))
+			item_image.copy_overlays(thing)
+		items += list("[thing.name] ([modkits_lenght])" = item_image)
```
```suggestion
		items["[thing.name] ([modkits_lenght])"] = item_image
```

Mothblocks commented on code/modules/projectiles/guns/energy/kinetic_accelerator.dm:
```diff
@@ -31,22 +31,62 @@
+
+	var/pick = show_radial_menu(user, src, items, custom_check = CALLBACK(src, PROC_REF(check_menu), user), radius = 36, require_near = TRUE, tooltips = TRUE)
+	if(!pick)
+		return
```
```suggestion
		return SECONDARY_ATTACK_CONTINUE_ATTACK_CHAIN
```

Mothblocks commented on code/modules/projectiles/guns/energy/kinetic_accelerator.dm:
```diff
@@ -31,22 +31,62 @@
+	var/modkit_reference = display_names[pick]
+	var/obj/item/borg/upgrade/modkit/modkit_to_remove = locate(modkit_reference) in modkits
+	if(!istype(modkit_to_remove))
+		return
```
```suggestion
		return SECONDARY_ATTACK_CONTINUE_ATTACK_CHAIN
```

Mothblocks reviewed (changes requested):
Add a contextual screentip as well

GuillaumePrata committed:
Implement Mothblock's requests

With a lil change to how drop_location worked as it was throwing errors for some reason

GuillaumePrata committed:
Swaps cancel for continue chain

GuillaumePrata committed:
Removes the work around to the `drop_location` issue

Because it was not needed, either VSC or I did a dummy

GuillaumePrata committed:
Revert the crowbar change
